[fix] Removed flattening due to slack webhook only accepts not flatte…

m5kr1pka opened the pull request:
###  Summary
Closes #11 Advanced messages fails issue, because slack only accepts not flattened parameters and payload doesn't reach workflow builder afterwards.

### Requirements

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/slack-github-action/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

li-olivierfostier commented:
Pleaaaase Hurry up ! :)

akasakashota commented:
Did you forget to build it? I'm looking forward to this PR being merged. Thanks.

seratch reviewed (approved):
LGTM but as I mentioned at https://github.com/slackapi/slack-github-action/issues/11#issuecomment-962688683, I would like to wait for @stevengill's review for this

seratch reviewed (changes requested):
Oops, I missed that the current implementation support only Workflow Builder webhook use cases. To support both webhooks for posting a message nad webhooks for triggering a workflow, perhaps, we need to:

* have a new option to determine the webhook type
* dispatch the logic accordingly (=the workflow webhook pattern should be supported too)

stevengill commented:
Agree with @seratch. I left a comment on the open issue. We don't want to remove flatten from the library but want to introduce a fork for incoming_webhook support.

m5kr1pka commented:
cc'ing:  @seratch @stevengill 

- Implemented a webhook type via env parameter: `SLACK_WEBHOOK_TYPE`: `BUILDER` or `INCOMING`, by default: `BUILDER` (existing);
- Updated documentation with Incoming webhook;
- added incoming webhook example to example-workflows;
- ran build script and committed unfortunately;
- added @vercel/ncc to devDependencies;

(phew)

P. S.
Would be great to implement tests in this action, but major release would be required

seratch commented on index.js:
```diff
@@ -4,16 +4,23 @@ const { WebClient } = require('@slack/web-api');
 const axios = require('axios');
 
+const SLACK_WEBHOOK_TYPES = {
+    WORKFLOW_BUILDER: 'BUILDER',
+    WORKFLOW_INCOMING: 'INCOMING'
```
As the "Incoming Webhooks" feature is not part of workflows / workflow builder, `WORKFLOW_INCOMING` does not make sense. Also, the value "BUILDER" can be confusing. How about renaming this way? Thoughts? @stevengill 
```suggestion
    WORKFLOW_TRIGGER: 'WORKFLOW_TRIGGER',
    INCOMING_WEBHOOK: 'INCOMING_WEBHOOK'
```

stevengill replied:
I agree @seratch

seratch commented on index.js:
```diff
@@ -4,16 +4,23 @@ const { WebClient } = require('@slack/web-api');
 try {
     const botToken = process.env.SLACK_BOT_TOKEN;
     const webhookUrl = process.env.SLACK_WEBHOOK_URL;
+    const webhookType = process.env.SLACK_WEBHOOK_TYPE?.toUpperCase() || SLACK_WEBHOOK_TYPES.WORKFLOW_BUILDER;
```
```suggestion
    // The default type is for Workflow Builder triggers. If you want to use this action for Incoming Webhooks, use the corresponding type instead.
    const webhookType = process.env.SLACK_WEBHOOK_TYPE?.toUpperCase() || SLACK_WEBHOOK_TYPES. WORKFLOW_TRIGGER;
```

m5kr1pka replied:
Note: whitespace in variable: `SLACK_WEBHOOK_TYPES. WORKFLOW_TRIGGER`

seratch replied:
Thanks, fixed.

seratch commented on README.md:
````diff
@@ -66,6 +66,31 @@ This route allows your GitHub Actions job to post a message in a Slack channel o
     SLACK_BOT_TOKEN: ${{ secrets.SLACK_BOT_TOKEN }}
 ```
 
+## Slack Incoming Webhook
````
Let's have "Route" for consistency

```suggestion
## Slack Incoming Webhook Route
```

seratch commented on README.md:
```diff
@@ -66,6 +66,31 @@ This route allows your GitHub Actions job to post a message in a Slack channel o
 
+## Slack Incoming Webhook
+
+This route allows your GitHub Actions job to post a message in a Slack channel or direct message by utilizing [messaging.webhooks](https://api.slack.com/messaging/webhooks) API method. With this route, you can instantly post a message.
```
```suggestion
This route allows your GitHub Actions job to post a message in a Slack channel or direct message by utilizing [Incoming Webhooks](https://api.slack.com/messaging/webhooks). With this route, you can instantly post a message.
```

seratch commented on README.md:
```diff
@@ -66,6 +66,31 @@ This route allows your GitHub Actions job to post a message in a Slack channel o
+## Setup
+
+* [Create a Slack App](https://api.slack.com/apps) for your workspace (alternatively use an existing app you have already created and installed)
+* Install the app to your workspace
+* Invite the bot user into the channel (`/invite @bot_user_name`)
```
```suggestion
* Add the [incoming-webhook](https://api.slack.com/scopes/incoming-webhook) bot scope under **OAuth & Permissions**
* Install the app to your workspace (you will select a channel to notify)
```

seratch commented on README.md:
```diff
@@ -66,6 +66,31 @@ This route allows your GitHub Actions job to post a message in a Slack channel o
+  with:
+    payload: "{\"blocks\":[{\"type\":\"section\",\"text\":{\"type\":\"mrkdwn\",\"text\":\"You have a new request: \"}},{\"type\":\"section\",\"fields\":[{\"type\":\"mrkdwn\",\"text\":\"*Type:* Computer (laptop)\"},{\"type\":\"mrkdwn\",\"text\":\"*When:* Submitted Aut 10\"}]}]}"
+  env:
+    SLACK_WEBHOOK_URL: ${{ secrets.SLACK_BOT_TOKEN }}
+    SLACK_WEBHOOK_TYPE: INCOMING
```
```suggestion
    SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
    SLACK_WEBHOOK_TYPE: INCOMING_WEBHOOK
```

seratch reviewed (changes requested):
A few comments. Please wait for @stevengill's review before starting to modify this PR

m5kr1pka commented:
@seratch changes look good to me, thanks for the review, waiting for confirmation.

stevengill commented on example-workflows/Incoming webhook/main.yml:
```diff
@@ -0,0 +1,16 @@
+        payload: "{\"text\":\"Danny Torrence left a 1 star review for your property.\",\"blocks\":[{\"type\":\"section\",\"text\":{\"type\":\"mrkdwn\",\"text\":\"Danny Torrence left the following review for your property:\"}},{\"type\":\"section\",\"block_id\":\"section567\",\"text\":{\"type\":\"mrkdwn\",\"text\":\"<https://example.com|Overlook Hotel> \\n :star: \\n Doors had too many axe holes,  ...
+      env:
+        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
+        SLACK_WEBHOOK_TYPE: incoming
```
```suggestion
        SLACK_WEBHOOK_TYPE: INCOMING_WEBHOOK
```

stevengill commented on index.js:
```diff
@@ -46,28 +54,37 @@ try {
 
-        // flatten JSON payload (no nested attributes)
-        const flatPayload = flatten(payload);
+        if (webhookType === SLACK_WEBHOOK_TYPES.WORKFLOW_BUILDER) {
```
```suggestion
        if (webhookType === SLACK_WEBHOOK_TYPES.WORKFLOW_TRIGGER) {
```

stevengill reviewed (approved):
LGTM
